optuna/storages/_journal/file.py

```diff
@@ -0,0 +1,175 @@
+            except BaseException:
+                self.release()
+                raise
+
+    def release(self) -> None:
+        try:
+            os.rename(self._lockfile, self._lockrenamefile)
+            os.unlink(self._lockrenamefile)
+        except OSError:
+            raise RuntimeError("Error: did not possess lock")
+        except BaseException:
+            os.unlink(self._lockrenamefile)
+            raise
+
+
+@experimental_class("3.1.0")
+class JournalFileOpenLock(JournalFileBaseLock):
+    """Lock class for synchronizing processes for NFSv3 or later.
+
+    On acquiring the lock, open system call is called with the O_EXCL option to create an exclusive
+    file. The file is deleted when the lock is released. This class is only supported when using
+    NFSv3 or later on kernel 2.6 or later. In prior NFS environments, use
+    :class:`~optuna.storages.JournalFileSymlinkLock`.
+
+    Args:
+        filepath:
+            The path of the file whose race condition must be protected.
+    """
+
+    def __init__(self, filepath: str) -> None:
+        self._lockfile = filepath + LOCK_FILE_SUFFIX
+        self._lockrenamefile = self._lockfile + str(uuid.uuid4()) + RENAME_FILE_SUFFIX
+
+    def acquire(self) -> bool:
+        while True:
+            try:
+                open_flags = os.O_CREAT | os.O_EXCL | os.O_WRONLY
+                os.close(os.open(self._lockfile, open_flags))
+                return True
+            except OSError as err:
+                if err.errno == errno.EEXIST:
+                    continue
+                else:
+                    raise err
+            except BaseException:
+                self.release()
+                raise
+
+    def release(self) -> None:
+        try:
+            os.rename(self._lockfile, self._lockrenamefile)
+            os.unlink(self._lockrenamefile)
+        except OSError:
+            raise RuntimeError("Error: did not possess lock")
+        except BaseException:
+            os.unlink(self._lockrenamefile)
+            raise
+
+
+@contextmanager
+def get_lock_file(lock_obj: JournalFileBaseLock) -> Iterator[None]:
+    lock_obj.acquire()
+    try:
+        yield
+    finally:
+        lock_obj.release()
+
+
+@experimental_class("3.1.0")
+class JournalFileStorage(BaseJournalLogStorage):
+    """File storage class for Journal log backend.
+
+    Args:
+        file_path:
+            Path of file to persist the log to.
+
+        lock_obj:
+            Lock object for process exclusivity.
+
+    """
+
+    def __init__(self, file_path: str, lock_obj: Optional[JournalFileBaseLock] = None) -> None:
+        self._file_path: str = file_path
+        self._lock = lock_obj or JournalFileSymlinkLock(self._file_path)
+        open(self._file_path, "a").close()  # Create a file if it does not exist
+        self._log_number_offset: Dict[int, int] = {0: 0}
+
+    def read_logs(self, log_number_from: int) -> List[Dict[str, Any]]:
+        with get_lock_file(self._lock):
```

Comment:
Do we actually need to acquire a lock here?
Since `JournalFileStorage` just appends JSONs to a file and doesn't modify the existing part of the file, it would be safe to read the contents of the file without lock (we should take care a case that the JSON at the file tail could be temporarily broken depending on the file write / flush timing though).
Let me know if I misunderstand something :pray:

Reply:
Nice catch!

As you mentioned, optimistic reading of the data entails risks of reading collapsed data due to concurrent writers, so we will need to add some mechanism that the data it read is not broken. Do you have any suggestions on doing so?
I think the easiest way to do so is to add a checksum for each line of JSON and persist it together. And when reading data, make sure that the checksum matches with the checksum computed from the line of JSON.

Reply:
IMO, it's enough to just ignore JSON decode error of the last line as follows:

```python
                last_decode_error = None # Added
                for log_number, line in enumerate(f, start=log_number_start):
                    if last_decode_error is not None:
                        raise last_decode_error # Added (the error is not at the last line)
                    if log_number + 1 not in self._log_number_offset:
                        byte_len = len(line.encode("utf-8"))
                        self._log_number_offset[log_number + 1] = (
                            self._log_number_offset[log_number] + byte_len
                        )
                    if log_number < log_number_from:
                        continue
                    try:        
                        logs.append(json.loads(line))
                    except e:
                        last_decode_err = e # Added
```

Reply:
Fixed in https://github.com/optuna/optuna/pull/3854/commits/d7869c0edb5f6b6a94d2f858ef65369563fb50ba